geom/twkb_write.go

```diff
@@ -0,0 +1,535 @@
+		w.writeRing(ls)
+	}
+	return nil
+}
+
+func (w *TWKBWriter) writeMultiPoint(mp MultiPoint) error {
+	w.writeTypeAndPrecision(twkbTypeMultiPoint)
+
+	if ctype := mp.CoordinatesType(); ctype != w.ctype {
+		return fmt.Errorf("mismatched coordinate dimensions got %s expected %s", ctype, w.ctype)
+	}
+
+	if mp.IsEmpty() {
+		w.writeIsEmptyHeader()
+		return nil
+	}
+	w.writeInitialHeaders()
+
+	numPoints := mp.NumPoints()
+	w.writeUnsignedVarint(uint64(numPoints))
+
+	if err := w.writeIDList(numPoints); err != nil {
+		return err
+	}
+
+	for i := 0; i < numPoints; i++ {
+		pt := mp.PointN(i)
+		w.writePointCoords(pt)
+	}
+	return nil
+}
+
+func (w *TWKBWriter) writeMultiLineString(ml MultiLineString) error {
+	w.writeTypeAndPrecision(twkbTypeMultiLineString)
+
+	if ctype := ml.CoordinatesType(); ctype != w.ctype {
+		return fmt.Errorf("mismatched coordinate dimensions got %s expected %s", ctype, w.ctype)
+	}
+
+	if ml.IsEmpty() {
+		w.writeIsEmptyHeader()
+		return nil
+	}
+	w.writeInitialHeaders()
+
+	numLineStrings := ml.NumLineStrings()
+	w.writeUnsignedVarint(uint64(numLineStrings))
+
+	if err := w.writeIDList(numLineStrings); err != nil {
+		return err
+	}
+
+	for i := 0; i < numLineStrings; i++ {
+		ls := ml.LineStringN(i)
+		w.writeLineStringCoords(ls)
+	}
+	return nil
+}
+
+func (w *TWKBWriter) writeMultiPolygon(mp MultiPolygon) error {
+	w.writeTypeAndPrecision(twkbTypeMultiPolygon)
+
+	if ctype := mp.CoordinatesType(); ctype != w.ctype {
+		return fmt.Errorf("mismatched coordinate dimensions got %s expected %s", ctype, w.ctype)
+	}
+
+	if mp.IsEmpty() {
+		w.writeIsEmptyHeader()
+		return nil
+	}
+	w.writeInitialHeaders()
+
+	numPolygons := mp.NumPolygons()
+	w.writeUnsignedVarint(uint64(numPolygons))
+
+	if err := w.writeIDList(numPolygons); err != nil {
+		return err
+	}
+
+	for i := 0; i < numPolygons; i++ {
+		poly := mp.PolygonN(i)
+		w.writePolygonRings(poly)
+	}
+	return nil
+}
+
+func (w *TWKBWriter) writeGeometryCollection(gc GeometryCollection) error {
+	w.writeTypeAndPrecision(twkbTypeGeometryCollection)
+
+	if ctype := gc.CoordinatesType(); ctype != w.ctype {
+		return fmt.Errorf("mismatched coordinate dimensions got %s expected %s", ctype, w.ctype)
+	}
+
+	if gc.IsEmpty() {
+		w.writeIsEmptyHeader()
+		return nil
+	}
+	w.writeInitialHeaders()
+
+	numGeometries := gc.NumGeometries()
+	w.writeUnsignedVarint(uint64(numGeometries))
+
+	if err := w.writeIDList(numGeometries); err != nil {
+		return err
+	}
+
+	for i := 0; i < numGeometries; i++ {
+		subWriter := copyTWKBWriter(w)
+		geom := gc.GeometryN(i)
+		subWriter.writeGeometry(geom)
+		w.twkb = append(w.twkb, subWriter.twkb...)
+	}
+	return nil
+}
+
+func (w *TWKBWriter) writeTypeAndPrecision(kind twkbGeometryType) {
+	w.kind = kind
+	w.writeByte(byte(EncodeZigZagInt32(int32(w.precXY))<<4) | byte(w.kind))
+}
+
+func (w *TWKBWriter) writeIsEmptyHeader() {
+	w.isEmpty = true
+	w.writeMetadataHeader(twkbIsEmpty)
+	// Do not write any extended info, bbox, size, or ids.
```

Comment:
Just checking: did you mean to place this comment here?

Reply:
That comment at the end of the function was to explain why there's nothing else to be done in that function (because it's an empty object so we stop right there). I'll rewrite that to make it clearer.